pkg/controllers/securitypolicy_controller.go

```diff
@@ -27,82 +29,108 @@ import (
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, METRIC_RES_TYPE)
-			return ctrl.Result{}, err
+			err = errors.New("fetch namespace associated with security policy CR failed")
+			log.Error(err, "would retry exponentially", "security-policy", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return resultRequeue, nil
 		} else if isCRInSysNs {
-			err = errors.New("Security Policy CR cannot be created in System Namespace")
-			log.Error(err, "failed to create security policy CR", "securitypolicy", req.NamespacedName)
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, METRIC_RES_TYPE)
-			return ctrl.Result{}, err
+			err = errors.New("security Policy CR cannot be created in System Namespace")
+			log.Error(err, "", "security-policy", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return resultNormal, nil
 		}
 
-		if err := r.Service.CreateOrUpdateSecurityPolicy(obj); err != nil {
-			log.Error(err, "failed to create or update security policy CR", "securitypolicy", req.NamespacedName)
-			r.setSecurityPolicyReadyStatusFalse(&ctx, obj, &err)
+		if err := r.Service.OperateSecurityPolicy(obj); err != nil {
+			log.Error(err, "operate failed, would retry exponentially", "security-policy", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
 			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, METRIC_RES_TYPE)
-			return ctrl.Result{}, err
+			return resultRequeue, nil
 		}
-		r.setSecurityPolicyReadyStatusTrue(&ctx, obj)
+		updateSuccess(r, &ctx, obj)
 		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateSuccessTotal, METRIC_RES_TYPE)
 	} else {
-		if containsString(obj.GetFinalizers(), util.FinalizerName) {
+		if controllerutil.ContainsFinalizer(obj, util.FinalizerName) {
 			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, METRIC_RES_TYPE)
 			if err := r.Service.DeleteSecurityPolicy(obj.UID); err != nil {
-				metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, METRIC_RES_TYPE)
-				return ctrl.Result{}, err
+				log.Error(err, "delete failed, would retry exponentially", "security-policy", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return resultRequeue, nil
 			}
 			controllerutil.RemoveFinalizer(obj, util.FinalizerName)
 			if err := r.Client.Update(ctx, obj); err != nil {
-				metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, METRIC_RES_TYPE)
-				return ctrl.Result{}, err
+				log.Error(err, "delete failed, would retry exponentially", "security-policy", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return resultRequeue, nil
 			}
-			log.V(1).Info("removed finalizer on securitypolicy CR", "securitypolicy", req.NamespacedName)
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, METRIC_RES_TYPE)
+			log.V(1).Info("removed finalizer", "security-policy", req.NamespacedName)
```

Comment:
I saw you change "securitypolicy" to "security-policy" in many places. I don't think it's a necessary change,  "securitypolicy" is enough because the customer uses `kubectl get securitypolicy`, not `kubectl get security-policy`. And NSX-T also has a kind of resource "security policy", so I think "securitypolicy" is better than "security-policy" to indicate that it's a CR, not NSX-T resource.